File: Data Structures & Algorithms/binary-tree-diameter/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    maxDiam = 0
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        self.dfs(root)
        return self.maxDiam

    def dfs(self, root: Optional[TreeNode]) -> int:
        if root is None:
            return 0
        
        diamLeft = self.dfs(root.left)
        diamRight = self.dfs(root.right)
        self.maxDiam = max(self.maxDiam, diamLeft+diamRight)

        return 1 + max(diamLeft, diamRight)

File: Data Structures & Algorithms/lowest-common-ancestor-in-binary-search-tree/submission-0.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        if (root.right == null && root.left == null){
            return root;
        }
        //if both lower than root
        if (p.val < root.val && q.val < root.val){
            return lowestCommonAncestor(root.left, p,q);
        }
        //if both greater than root
        else if(p.val > root.val && q.val > root.val){
            return lowestCommonAncestor(root.right,p,q);
        }
        //if split by root then root
        return root;
    }
}
